src/Aspire.Hosting/ApplicationModel/ResourceCommandAnnotation.cs

```diff
@@ -0,0 +1,99 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace Aspire.Hosting.ApplicationModel;
+
+#pragma warning disable RS0016 // Add public types and members to the declared API
+/// <summary>
+/// Represents a command annotation for a resource.
+/// </summary>
+[DebuggerDisplay("Type = {GetType().Name,nq}, Type = {Type}")]
+public sealed class ResourceCommandAnnotation : IResourceAnnotation
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResourceCommandAnnotation"/> class.
+    /// </summary>
+    public ResourceCommandAnnotation(
+        string type,
+        string displayName,
+        Func<UpdateCommandStateContext, ResourceCommandState> updateState,
+        Func<ExecuteCommandContext, Task> executeCommand,
```

Comment:
I was thinking about whether we should return something like the equivalent of `Task<IActionResult>` (not actual `IActionResult` just trying to conjure the pattern). The idea is that that no only does the task being awaited some indication of success or failure, but it also allows us to provide the dashboard with additional information that can be used to control what happens.

For example we might want to provide some kind of call to action button that the user can click to navigate away to a page - or in some cases a redirect action which would immediately navigate the page away to a specific URL. For example if someone adds the `WithPgAdmin(...)` command then all the databases on that server would get a command _View in PgAdmin_ that if they clicked on it would be deep linked to that database in the PgAdmin instance that the database is mapped to.

Other options for projects might be doing things like scanning the project directory for .http files and adding a command for each .http file so that when it is clicked the .http request is executed and the JSON result is displayed in a JSON viewer in the screen.

We don't necessarily need all these options straightaway but given that there are some useful scenarios here making this return a more fully fledged object might be an idea. Then we could extend it overtime with more scenarios.